### PR TITLE
Add release task

### PR DESCRIPTION
**Pull request by matt-gadd**
This attempts to automate away some of the `publishing` part of https://github.com/dojo/meta/blob/master/documents/Committer-Workflow.md

The task will:
* check that the branch is clean, and on the `master` branch
* check the user is both logged into npm, and has permissions to publish the package
* select the correct pre tag release version based on what is in the npm registry
* update the `package.json`, commit and tag
* prepare the package as a flat distribution
* publish to npm
* update the `package.json` version and commit

(it does **not** push back to the remote) 

for now, if wanting to bump the major, minor or patch on semver - you have to manually specify the `release-version` and `next-version` appropriately (this could be deduced in the future).

for pre release tags the version is set back to whatever pre version it was before automatically.

to release an alpha tag for example:
`grunt release --pre-release-tag=alpha`

to release a specific version: 
`grunt release --release-version=3.0.1 --next-version=3.0.2-pre`

you can specify an npm tag:
`grunt release --pre-release-tag=beta --tag=next`

to dry run:
`grunt release --pre-release-tag=alpha --dry-run`

`--dry-run` will modify the `package.json`, but not make any commits or tags, and rather than publishing to npm, it will pack it so you can inspect what the distribution would be.

### CODE REVIEW

File: tasks/release.ts [hunk 1]
```diff
@@ -0,0 +1,143 @@
...
+		const nextVersion = filteredVersions.length ? Math.max(...filteredVersions) + 1 : 1;
+		return `${versionInPackage}-${preReleaseTag}.${nextVersion}`;
+	}
+
+	function preparePackageJson(packageJson: any): any {
+		packageJson.private = undefined;
+		packageJson.scripts = undefined;
+		packageJson.files = undefined;
+		packageJson.typings = undefined;
+		packageJson.main = 'main.js';
```
Comment (matt-gadd) [inline tasks/release.ts]: some of this can disappear when we fix up some of our package.json's

File: tasks/release.ts [hunk 2]
```diff
@@ -0,0 +1,143 @@
+export = function(grunt: IGrunt, packageJson: any) {
+	const execa = require('execa');
+	const npmPath = 'node_modules/.bin/npm';
+	const temp = 'temp/';
+	const preReleaseTags = ['alpha', 'beta', 'rc'];
+
+	const releaseVersion = grunt.option<string>('release-version');
```
Comment (kitsonk) [inline tasks/release.ts]: you could do:

```typescript
const releaseVersion: string = grunt.option('release-version');
```

It reads a bit better to me, but I am open to alternative viewpoints

**Comment by kitsonk** [comment]
Looking good... I like that fact that `grunt release` is a noop, because that is too dangerous.

How does it deal with logged in user in `npm`?

**Comment by matt-gadd** [comment]
@kitsonk it currently doesn't deal with anything on the logging in side. so the appropriate user would need to be logged in, otherwise the publish would fail (although you can still dry run obviously). I think I probably need to take a look at conveying some friendly error messages for those scenario's.

**Comment by matt-gadd** [comment]
perhaps i can do a `whoami` and check whether the user aligns with the package's maintainers

File: tasks/release.ts [hunk 3]
```diff
@@ -0,0 +1,188 @@
...
+	grunt.registerTask('release-publish', 'publish the package to npm', function () {
+		const done = this.async();
+		const args = ['publish', '.'];
+		const promises = [command(npmBin, args, { cwd: temp }, false)];
+		if (tag) {
+			args.push('--tag', tag);
+		}
+		grunt.log.subhead('publishing to npm...');
+		if (dryRun) {
+			promises.push(command(npmBin, ['pack', '../temp'], { cwd: 'dist' }, true));
```
Comment (matt-gadd) [inline tasks/release.ts]: use the `temp` const and get the `dist` dir from config

**Comment by agubler** [comment]
👍

**Comment by matt-gadd** [comment]
I think this is good to go. I will raise a follow up PR which will restore `npm link` functionality via a grunt task.

**Comment by agubler** [comment]
👍 (again), @kitsonk are you happy for this to go in?

**Comment by kitsonk** [comment]
👍

**Pull request merged**